<a href="https://colab.research.google.com/github/Tawanss/BSC_DPDM23/blob/main/Frequent_Patterns_(Association_Rules).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [13]:
import pandas as pd
import os

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [16]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/MyDrive/data minning/reduced_marketbasket.csv'

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [17]:
data = pd.read_csv(os.path.join(path,'/content/drive/MyDrive/data minning/reduced_marketbasket.csv'),encoding='latin-1')

In [18]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

Quiz: ข้อมูลมาจากกี่ประเทศ

In [ ]:
len(set(data['Country']))

38

In [ ]:
data['Country'].unique()

array(['United Kingdom', 'France', 'Australia', 'Netherlands', 'Germany',
       'Norway', 'EIRE', 'Switzerland', 'Spain', 'Poland', 'Portugal',
       'Italy', 'Belgium', 'Lithuania', 'Japan', 'Iceland',
       'Channel Islands', 'Denmark', 'Cyprus', 'Sweden', 'Austria',
       'Israel', 'Finland', 'Bahrain', 'Greece', 'Hong Kong', 'Singapore',
       'Lebanon', 'United Arab Emirates', 'Saudi Arabia',
       'Czech Republic', 'Canada', 'Unspecified', 'Brazil', 'USA',
       'European Community', 'Malta', 'RSA'], dtype=object)

# แปลง data ให้อยู่ในรูปแบบ Transaction data ด้วย *groupby.apply(list)*

In [ ]:
data.groupby('InvoiceNo')['StockCode'].apply(list)

InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
C581484                                              [23843]
C581490                                       [22178, 23144]
C581499                                                  [M]
C581568                                              [21258]
C581569                                       [84978, 20979]
Name: StockCode, Length: 25900, dtype: object

In [ ]:
transactions = data.groupby('InvoiceNo')['StockCode'].apply(list)

## Association Rules

In [19]:
!pip install apyori

  Preparing metadata (setup.py) ... done
  Created wheel for apyori: filename=apyori-1.1.2-py3-none-any.whl size=5954 sha256=8c21bba51c8fa3a25c7d661701320da0cb758c7b9813451750f21b3ddecb15d3
  Stored in directory: /root/.cache/pip/wheels/c4/1a/79/20f55c470a50bb3702a8cb7c94d8ada15573538c7f4baebe2d
Successfully built apyori


In [20]:
from apyori import apriori

In [ ]:
apriori(transactions,min_support = 0.5)

<generator object apriori at 0x7a33de7f7ca0>

In [ ]:
list(apriori(transactions,min_support = 0.065))

[RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)]),
 RelationRecord(items=frozenset({'47566'}), support=0.06586872586872587, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'47566'}), confidence=0.06586872586872587, lift=1.0)]),
 RelationRecord(items=frozenset({'85099B'}), support=0.08243243243243244, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.08243243243243244, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.08671814671814672, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08671814671814672, lift=1.0)])]

In [ ]:
frequentItemsets = list(apriori(transactions,min_support = 0.065))
frequentItemsets

[RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)]),
 RelationRecord(items=frozenset({'47566'}), support=0.06586872586872587, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'47566'}), confidence=0.06586872586872587, lift=1.0)]),
 RelationRecord(items=frozenset({'85099B'}), support=0.08243243243243244, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.08243243243243244, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.08671814671814672, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08671814671814672, lift=1.0)])]

In [ ]:
len(frequentItemsets)

4

In [ ]:
frequentItemsets[0]

RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)])

In [ ]:
frequentItemsets[0][0]

frozenset({'22423'})

In [ ]:
dir(frequentItemsets[0][0])

['__and__',
 '__class__',
 '__class_getitem__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__ne__',
 '__new__',
 '__or__',
 '__rand__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__ror__',
 '__rsub__',
 '__rxor__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__sub__',
 '__subclasshook__',
 '__xor__',
 'copy',
 'difference',
 'intersection',
 'isdisjoint',
 'issubset',
 'issuperset',
 'symmetric_difference',
 'union']

In [ ]:
frequentItemsets[0][0].__len__()

1

In [ ]:
min_sup = 0.03

frequentItemsets = list(apriori(transactions,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'85099B', '22386'}), support=0.03216216216216216, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B', '22386'}), confidence=0.03216216216216216, lift=1.0), OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])
yes
RelationRecord(items=frozenset({'22697', '22699'}), support=0.03027027027027027, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22697', '22699'}), confidence=0.03027027027027027, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])


RelationRecord(items=frozenset({'22386', '85099B'}),

support=0.03216216216216216,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0),

OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335),

OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])


In [ ]:
set(data[(data['StockCode']=='22386') | (data['StockCode']=='85099B') | (data['StockCode']=='22699') | (data['StockCode']=='22697') ]['Description'])

{'GREEN REGENCY TEACUP AND SAUCER',
 'JUMBO BAG PINK POLKADOT',
 'JUMBO BAG RED RETROSPOT',
 'ROSES REGENCY TEACUP AND SAUCER '}

Quiz <br>
stock code 22386 and 85099B คือสินค้าอะไร

In [ ]:
set(data[(data['StockCode']=='22386')]['Description'])

{'JUMBO BAG PINK POLKADOT'}

In [ ]:
set(data[(data['StockCode']=='85099B')]['Description'])

{'JUMBO BAG RED RETROSPOT'}

RelationRecord(items=frozenset({'22699', '22697'}), support=0.03027027027027027,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.03027027027027027, lift=1.0),
<br>
<br>

OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947),

OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])

In [ ]:
set(data[(data['StockCode']=='22699') | (data['StockCode']=='22697')]['InvoiceNo'])

{'552694',
 '539442',
 '538071',
 '558030',
 '570465',
 '566624',
 '577594',
 '552804',
 '549524',
 'C548485',
 '558565',
 '575185',
 'C558434',
 '547493',
 'C541822',
 'C556644',
 '544089',
 '557642',
 '566602',
 '552247',
 '578065',
 '550474',
 '554030',
 '556243',
 '547846',
 'C543633',
 '562534',
 '554113',
 '545377',
 '553578',
 '545320',
 '553607',
 '562607',
 '570720',
 '546752',
 '565155',
 '557605',
 '542349',
 '550307',
 '559930',
 '551717',
 '575138',
 '551286',
 '542632',
 '554643',
 '566788',
 '541932',
 '568048',
 '550953',
 '541132',
 'C538124',
 '556095',
 '550500',
 '550210',
 '565919',
 'C546125',
 '537754',
 '564820',
 '579445',
 '545912',
 '557621',
 '543532',
 '547387',
 '559491',
 '562353',
 '572758',
 '544778',
 '565424',
 '566617',
 '558119',
 '575206',
 '546688',
 '554040',
 '576672',
 '542629',
 '552642',
 '553013',
 '546212',
 '551340',
 '561798',
 'C556880',
 '558240',
 '547063',
 '548734',
 '558372',
 '570700',
 '549999',
 '540465',
 '554382',
 'C569114',
 

In [ ]:
data[data['InvoiceNo']=='562950']

InvoiceNo StockCode                         Description  Quantity  \
297488    562950     22752        SET 7 BABUSHKA NESTING BOXES         2   
297489    562950     22755     SMALL PURPLE BABUSHKA NOTEBOOK         12   
297490    562950     22753     SMALL YELLOW BABUSHKA NOTEBOOK         12   
297491    562950     22754        SMALL RED BABUSHKA NOTEBOOK         12   
297492    562950     22757        LARGE RED BABUSHKA NOTEBOOK         12   
297493    562950     22758    LARGE PURPLE BABUSHKA NOTEBOOK          12   
297494    562950     22756     LARGE YELLOW BABUSHKA NOTEBOOK         12   
297495    562950     23310             BUBBLEGUM RING ASSORTED        36   
297496    562950     23280  FOLDING BUTTERFLY MIRROR HOT PINK         12   
297497    562950     22383              LUNCH BAG SUKI DESIGN         10   
297498    562950     22492             MINI PAINT SET VINTAGE         36   
297499    562950     71459      HANGING JAM JAR T-LIGHT HOLDER        12   
297500    562950     84945  MULTI COLOUR SILVER T-LIGHT HOLDER        12   
297501    562950     21175         GIN + TONIC DIET METAL SIGN        12   
297502    562950     23164                  REGENCY CAKE SLICE         4   
297503    562950     23245          SET OF 3 REGENCY CAKE TINS         4   
297504    562950     23171            REGENCY TEA PLATE GREEN         12   
297505    562950     22697     GREEN REGENCY TEACUP AND SAUCER         6   

             InvoiceDate  UnitPrice  CustomerID         Country  
297488  08/11/2011 09:56       8.50     15608.0  United Kingdom  
297489  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297490  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297491  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297492  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297493  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297494  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297495  08/11/2011 09:56       0.42     15608.0  United Kingdom  
297496  08/11/2011 09:56       0.83     15608.0  United Kingdom  
297497  08/11/2011 09:56       1.65     15608.0  United Kingdom  
297498  08/11/2011 09:56       0.65     15608.0  United Kingdom  
297499  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297500  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297501  08/11/2011 09:56       2.55     15608.0  United Kingdom  
297502  08/11/2011 09:56       4.95     15608.0  United Kingdom  
297503  08/11/2011 09:56       4.95     15608.0  United Kingdom  
297504  08/11/2011 09:56       1.65     15608.0  United Kingdom  
297505  08/11/2011 09:56       2.95     15608.0  United Kingdom

#HW 1 หา Frequent Item Set และ Association Rules
* ของช่วงเวลา Christmas & New Year (1 Dec - 31 Dec)
* ของประเทศในทวีป Europe
* ของช่วงเวลา ตรุษจีน ในทวีป Asia  (15 Jan - 15 Feb)
* ของช่วงเวลา  Christmas & New Year ในทวีป Europe (1 Dec - 31 Dec)
* ของวันวาเลนไทน์
* โดยจัดกลุ่มตามประเทศ + เดือน


In [ ]:
data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

        InvoiceDate  UnitPrice  CustomerID         Country  
0  12/01/2010 08:26       2.55     17850.0  United Kingdom  
1  12/01/2010 08:26       3.39     17850.0  United Kingdom  
2  12/01/2010 08:26       2.75     17850.0  United Kingdom  
3  12/01/2010 08:26       3.39     17850.0  United Kingdom  
4  12/01/2010 08:26       3.39     17850.0  United Kingdom

In [21]:
data_HW = data.copy()
data_HW

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

In [22]:

# Given data
countries = ['United Kingdom', 'France', 'Australia', 'Netherlands', 'Germany',
             'Norway', 'EIRE', 'Switzerland', 'Spain', 'Poland', 'Portugal',
             'Italy', 'Belgium', 'Lithuania', 'Japan', 'Iceland',
             'Channel Islands', 'Denmark', 'Cyprus', 'Sweden', 'Austria',
             'Israel', 'Finland', 'Bahrain', 'Greece', 'Hong Kong', 'Singapore',
             'Lebanon', 'United Arab Emirates', 'Saudi Arabia',
             'Czech Republic', 'Canada', 'Unspecified', 'Brazil', 'USA',
             'European Community', 'Malta', 'RSA']

# Fictional mapping of countries to continents (replace with accurate data)
continent_mapping = {
    'United Kingdom': 'Europe',
    'France': 'Europe',
    'Australia': 'Oceania',
    'Netherlands': 'Europe',
    'Germany': 'Europe',
    'Norway': 'Europe',
    'EIRE': 'Europe',
    'Switzerland': 'Europe',
    'Spain': 'Europe',
    'Poland': 'Europe',
    'Portugal': 'Europe',
    'Italy': 'Europe',
    'Belgium': 'Europe',
    'Lithuania': 'Europe',
    'Japan': 'Asia',
    'Iceland': 'Europe',
    'Channel Islands': 'Europe',
    'Denmark': 'Europe',
    'Cyprus': 'Europe',
    'Sweden': 'Europe',
    'Austria': 'Europe',
    'Israel': 'Asia',
    'Finland': 'Europe',
    'Bahrain': 'Asia',
    'Greece': 'Europe',
    'Hong Kong': 'Asia',
    'Singapore': 'Asia',
    'Lebanon': 'Asia',
    'United Arab Emirates': 'Asia',
    'Saudi Arabia': 'Asia',
    'Czech Republic': 'Europe',
    'Canada': 'North America',
    'Unspecified': 'Unknown',
    'Brazil': 'South America',
    'USA': 'North America',
    'European Community': 'Europe',
    'Malta': 'Europe',
    'RSA': 'Africa'
}

# Create a DataFrame
data_continent = {'Country': countries, 'Continent': [continent_mapping.get(country, 'Unknown') for country in countries]}
df_continent = pd.DataFrame(data_continent)

# Display the DataFrame
df_continent.head()


Country Continent
0  United Kingdom    Europe
1          France    Europe
2       Australia   Oceania
3     Netherlands    Europe
4         Germany    Europe

In [23]:
df_continent

Country      Continent
0         United Kingdom         Europe
1                 France         Europe
2              Australia        Oceania
3            Netherlands         Europe
4                Germany         Europe
5                 Norway         Europe
6                   EIRE         Europe
7            Switzerland         Europe
8                  Spain         Europe
9                 Poland         Europe
10              Portugal         Europe
11                 Italy         Europe
12               Belgium         Europe
13             Lithuania         Europe
14                 Japan           Asia
15               Iceland         Europe
16       Channel Islands         Europe
17               Denmark         Europe
18                Cyprus         Europe
19                Sweden         Europe
20               Austria         Europe
21                Israel           Asia
22               Finland         Europe
23               Bahrain           Asia
24                Greece         Europe
25             Hong Kong           Asia
26             Singapore           Asia
27               Lebanon           Asia
28  United Arab Emirates           Asia
29          Saudi Arabia           Asia
30        Czech Republic         Europe
31                Canada  North America
32           Unspecified        Unknown
33                Brazil  South America
34                   USA  North America
35    European Community         Europe
36                 Malta         Europe
37                   RSA         Africa

In [24]:
df_continent['Continent'].value_counts()

Europe           24
Asia              8
North America     2
Oceania           1
Unknown           1
South America     1
Africa            1
Name: Continent, dtype: int64

In [25]:
df_continent['Continent'].count()

38

# เช็คประเทศ

In [26]:
df_continent.sort_values(by = 'Continent')

Country      Continent
37                   RSA         Africa
14                 Japan           Asia
29          Saudi Arabia           Asia
28  United Arab Emirates           Asia
27               Lebanon           Asia
26             Singapore           Asia
25             Hong Kong           Asia
23               Bahrain           Asia
21                Israel           Asia
19                Sweden         Europe
20               Austria         Europe
24                Greece         Europe
36                 Malta         Europe
30        Czech Republic         Europe
35    European Community         Europe
22               Finland         Europe
17               Denmark         Europe
0         United Kingdom         Europe
15               Iceland         Europe
1                 France         Europe
3            Netherlands         Europe
4                Germany         Europe
16       Channel Islands         Europe
6                   EIRE         Europe
7            Switzerland         Europe
5                 Norway         Europe
9                 Poland         Europe
10              Portugal         Europe
11                 Italy         Europe
12               Belgium         Europe
13             Lithuania         Europe
8                  Spain         Europe
18                Cyprus         Europe
31                Canada  North America
34                   USA  North America
2              Australia        Oceania
33                Brazil  South America
32           Unspecified        Unknown

In [27]:
data_HW_continent = data_HW.merge(df_continent, how ='left', left_on='Country',right_on='Country')
print(data_HW_continent.shape)
data_HW_continent.head()

(541909, 9)


InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

        InvoiceDate  UnitPrice  CustomerID         Country Continent  
0  12/01/2010 08:26       2.55     17850.0  United Kingdom    Europe  
1  12/01/2010 08:26       3.39     17850.0  United Kingdom    Europe  
2  12/01/2010 08:26       2.75     17850.0  United Kingdom    Europe  
3  12/01/2010 08:26       3.39     17850.0  United Kingdom    Europe  
4  12/01/2010 08:26       3.39     17850.0  United Kingdom    Europe

In [28]:
data_HW_continent

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country Continent  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom    Europe  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom    Europe  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom    Europe  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom    Europe  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom    Europe  
...                  ...        ...         ...             ...       ...  
541904  12/09/2011 12:50       0.85     12680.0          France    Europe  
541905  12/09/2011 12:50       2.10     12680.0          France    Europe  
541906  12/09/2011 12:50       4.15     12680.0          France    Europe  
541907  12/09/2011 12:50       4.15     12680.0          France    Europe  
541908  12/09/2011 12:50       4.95     12680.0          France    Europe  

[541909 rows x 9 columns]

In [29]:
data_HW_continent_date = data_HW_continent.copy()

In [30]:
data_HW_continent_date['date_time'] = pd.to_datetime(data_HW_continent_date['InvoiceDate'])

In [31]:
data_HW_continent_date.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

        InvoiceDate  UnitPrice  CustomerID         Country Continent  \
0  12/01/2010 08:26       2.55     17850.0  United Kingdom    Europe   
1  12/01/2010 08:26       3.39     17850.0  United Kingdom    Europe   
2  12/01/2010 08:26       2.75     17850.0  United Kingdom    Europe   
3  12/01/2010 08:26       3.39     17850.0  United Kingdom    Europe   
4  12/01/2010 08:26       3.39     17850.0  United Kingdom    Europe   

            date_time  
0 2010-12-01 08:26:00  
1 2010-12-01 08:26:00  
2 2010-12-01 08:26:00  
3 2010-12-01 08:26:00  
4 2010-12-01 08:26:00

In [32]:
data_HW_continent_date.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 541909 entries, 0 to 541908
Data columns (total 10 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  object        
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
 8   Continent    541909 non-null  object        
 9   date_time    541909 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(2), int64(1), object(6)
memory usage: 45.5+ MB


# Data Prep

In [33]:
data.isnull().any()

InvoiceNo      False
StockCode      False
Description     True
Quantity       False
InvoiceDate    False
UnitPrice      False
CustomerID      True
Country        False
dtype: bool

In [34]:
data[data['Description'].isnull()]

InvoiceNo StockCode Description  Quantity       InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/01/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/01/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/01/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/01/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/01/2010 14:34        0.0   
...          ...       ...         ...       ...               ...        ...   
535322    581199     84581         NaN        -2  12/07/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/07/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/07/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/08/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/08/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

In [35]:
len(set(data[data['Description'].isnull()]['StockCode']))

960

In [36]:
data[data['StockCode'] == '22139']

InvoiceNo StockCode                       Description  Quantity  \
106       536381     22139  RETROSPOT TEA SET CERAMIC 11 PC         23   
622       536414     22139                               NaN        56   
6392      536942     22139                            amazon        15   
6885      536982     22139  RETROSPOT TEA SET CERAMIC 11 PC         10   
7203      537011     22139                               NaN        -5   
...          ...       ...                               ...       ...   
538411    581405     22139  RETROSPOT TEA SET CERAMIC 11 PC          1   
539531    581439     22139  RETROSPOT TEA SET CERAMIC 11 PC          1   
540441    581486     22139  RETROSPOT TEA SET CERAMIC 11 PC          6   
541387    581498     22139  RETROSPOT TEA SET CERAMIC 11 PC          2   
541726    581570     22139  RETROSPOT TEA SET CERAMIC 11 PC          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
106     12/01/2010 09:41       4.25     15311.0  United Kingdom  
622     12/01/2010 11:52       0.00         NaN  United Kingdom  
6392    12/03/2010 12:08       0.00         NaN  United Kingdom  
6885    12/03/2010 14:27      11.02         NaN  United Kingdom  
7203    12/03/2010 15:38       0.00         NaN  United Kingdom  
...                  ...        ...         ...             ...  
538411  12/08/2011 13:50       4.95     13521.0  United Kingdom  
539531  12/08/2011 16:30      10.79         NaN  United Kingdom  
540441  12/09/2011 09:38       4.95     17001.0  United Kingdom  
541387  12/09/2011 10:26      10.79         NaN  United Kingdom  
541726  12/09/2011 11:59       4.95     12662.0         Germany  

[993 rows x 8 columns]

In [37]:
data[data['UnitPrice'] <0 ]

InvoiceNo StockCode      Description  Quantity       InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  08/12/2011 14:51   
299984   A563187         B  Adjust bad debt         1  08/12/2011 14:52   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

In [38]:
data.iloc[299980:299986]

InvoiceNo StockCode                        Description  Quantity  \
299980    563184    82494L        WOODEN FRAME ANTIQUE WHITE          4   
299981    563184     82482  WOODEN PICTURE FRAME WHITE FINISH         4   
299982   A563185         B                    Adjust bad debt         1   
299983   A563186         B                    Adjust bad debt         1   
299984   A563187         B                    Adjust bad debt         1   
299985    563188     79160      HEART SHAPE WIRELESS DOORBELL        48   

             InvoiceDate  UnitPrice  CustomerID         Country  
299980  08/12/2011 14:50       2.95     17516.0  United Kingdom  
299981  08/12/2011 14:50       2.55     17516.0  United Kingdom  
299982  08/12/2011 14:50   11062.06         NaN  United Kingdom  
299983  08/12/2011 14:51  -11062.06         NaN  United Kingdom  
299984  08/12/2011 14:52  -11062.06         NaN  United Kingdom  
299985  08/12/2011 15:00       1.69     15606.0  United Kingdom

In [39]:
data[data['StockCode'] == 'B']

InvoiceNo StockCode      Description  Quantity       InvoiceDate  \
299982   A563185         B  Adjust bad debt         1  08/12/2011 14:50   
299983   A563186         B  Adjust bad debt         1  08/12/2011 14:51   
299984   A563187         B  Adjust bad debt         1  08/12/2011 14:52   

        UnitPrice  CustomerID         Country  
299982   11062.06         NaN  United Kingdom  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

In [40]:
data[data['InvoiceNo'] == 'A563187' ]

InvoiceNo StockCode      Description  Quantity       InvoiceDate  \
299984   A563187         B  Adjust bad debt         1  08/12/2011 14:52   

        UnitPrice  CustomerID         Country  
299984  -11062.06         NaN  United Kingdom

In [41]:
data[data['Description'].isnull()]['StockCode'].value_counts()

35965     10
23084     10
22084      9
22451      6
22501      5
          ..
16169P     1
84846A     1
82545A     1
84251F     1
85175      1
Name: StockCode, Length: 960, dtype: int64

In [42]:
data[data['Quantity']<0]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

             InvoiceDate  UnitPrice  CustomerID         Country  
141     12/01/2010 09:41      27.50     14527.0  United Kingdom  
154     12/01/2010 09:49       4.65     15311.0  United Kingdom  
235     12/01/2010 10:24       1.65     17548.0  United Kingdom  
236     12/01/2010 10:24       0.29     17548.0  United Kingdom  
237     12/01/2010 10:24       0.29     17548.0  United Kingdom  
...                  ...        ...         ...             ...  
540449  12/09/2011 09:57       0.83     14397.0  United Kingdom  
541541  12/09/2011 10:28     224.69     15498.0  United Kingdom  
541715  12/09/2011 11:57      10.95     15311.0  United Kingdom  
541716  12/09/2011 11:58       1.25     17315.0  United Kingdom  
541717  12/09/2011 11:58       1.25     17315.0  United Kingdom  

[10624 rows x 8 columns]

In [43]:
data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

In [44]:
data_positive = data[(data['Quantity']>=0) & (data['UnitPrice'] >=0) ]
data_positive

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[531283 rows x 8 columns]

In [45]:
data_positive.describe()

Quantity      UnitPrice     CustomerID
count  531283.000000  531283.000000  397924.000000
mean       10.655299       3.898954   15294.315171
std       156.830617      35.876279    1713.169877
min         1.000000       0.000000   12346.000000
25%         1.000000       1.250000   13969.000000
50%         3.000000       2.080000   15159.000000
75%        10.000000       4.130000   16795.000000
max     80995.000000   13541.330000   18287.000000

In [46]:
data_positive[data_positive['UnitPrice'] == 0]

InvoiceNo StockCode                   Description  Quantity  \
622       536414     22139                           NaN        56   
1970      536545     21134                           NaN         1   
1971      536546     22145                           NaN         1   
1972      536547     37509                           NaN         1   
1987      536549    85226A                           NaN         1   
...          ...       ...                           ...       ...   
535334    581211     22142                         check        14   
536981    581234     72817                           NaN        27   
538504    581406    46000M  POLYESTER FILLER PAD 45x45cm       240   
538505    581406    46000S  POLYESTER FILLER PAD 40x40cm       300   
538554    581408     85175                           NaN        20   

             InvoiceDate  UnitPrice  CustomerID         Country  
622     12/01/2010 11:52        0.0         NaN  United Kingdom  
1970    12/01/2010 14:32        0.0         NaN  United Kingdom  
1971    12/01/2010 14:33        0.0         NaN  United Kingdom  
1972    12/01/2010 14:33        0.0         NaN  United Kingdom  
1987    12/01/2010 14:34        0.0         NaN  United Kingdom  
...                  ...        ...         ...             ...  
535334  12/07/2011 18:36        0.0         NaN  United Kingdom  
536981  12/08/2011 10:33        0.0         NaN  United Kingdom  
538504  12/08/2011 13:58        0.0         NaN  United Kingdom  
538505  12/08/2011 13:58        0.0         NaN  United Kingdom  
538554  12/08/2011 14:06        0.0         NaN  United Kingdom  

[1179 rows x 8 columns]

In [47]:
data[(data['InvoiceNo'] == 'C581490') | (data['InvoiceNo'] == 'C581499')]

InvoiceNo StockCode                      Description  Quantity  \
540448   C581490     22178  VICTORIAN GLASS HANGING T-LIGHT       -12   
540449   C581490     23144  ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                           Manual        -1   

             InvoiceDate  UnitPrice  CustomerID         Country  
540448  12/09/2011 09:57       1.95     14397.0  United Kingdom  
540449  12/09/2011 09:57       0.83     14397.0  United Kingdom  
541541  12/09/2011 10:28     224.69     15498.0  United Kingdom

In [48]:
data[(data['CustomerID'] == 14397)]

InvoiceNo StockCode                        Description  Quantity  \
182820    552575     22178    VICTORIAN GLASS HANGING T-LIGHT        24   
182821    552575     23144    ZINC T-LIGHT HOLDER STARS SMALL        12   
182822    552575     20846  ZINC HEART LATTICE T-LIGHT HOLDER        12   
182823    552575     82484  WOOD BLACK BOARD ANT WHITE FINISH         2   
182824    552575     84946       ANTIQUE SILVER T-LIGHT GLASS        12   
...          ...       ...                                ...       ...   
533101    581011     22478           BIRDHOUSE GARDEN MARKER         20   
533102    581011     20846  ZINC HEART LATTICE T-LIGHT HOLDER        24   
533103    581011     22600      CHRISTMAS RETROSPOT STAR WOOD        12   
540448   C581490     22178    VICTORIAN GLASS HANGING T-LIGHT       -12   
540449   C581490     23144    ZINC T-LIGHT HOLDER STARS SMALL       -11   

             InvoiceDate  UnitPrice  CustomerID         Country  
182820  05/10/2011 12:25       1.25     14397.0  United Kingdom  
182821  05/10/2011 12:25       0.83     14397.0  United Kingdom  
182822  05/10/2011 12:25       1.25     14397.0  United Kingdom  
182823  05/10/2011 12:25       7.95     14397.0  United Kingdom  
182824  05/10/2011 12:25       1.25     14397.0  United Kingdom  
...                  ...        ...         ...             ...  
533101  12/07/2011 09:22       1.25     14397.0  United Kingdom  
533102  12/07/2011 09:22       1.25     14397.0  United Kingdom  
533103  12/07/2011 09:22       0.85     14397.0  United Kingdom  
540448  12/09/2011 09:57       1.95     14397.0  United Kingdom  
540449  12/09/2011 09:57       0.83     14397.0  United Kingdom  

[95 rows x 8 columns]

## ตัดรายการที่มีค่า Quantity เป็น ลบ และ UnitPrice < 0

In [49]:
data_HW_continent_date_clean1 = data_HW_continent_date[(data_HW_continent_date['Quantity']>=0) & (data_HW_continent_date['UnitPrice'] >=0) ]
data_HW_continent_date_clean1

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country Continent  \
0       12/01/2010 08:26       2.55     17850.0  United Kingdom    Europe   
1       12/01/2010 08:26       3.39     17850.0  United Kingdom    Europe   
2       12/01/2010 08:26       2.75     17850.0  United Kingdom    Europe   
3       12/01/2010 08:26       3.39     17850.0  United Kingdom    Europe   
4       12/01/2010 08:26       3.39     17850.0  United Kingdom    Europe   
...                  ...        ...         ...             ...       ...   
541904  12/09/2011 12:50       0.85     12680.0          France    Europe   
541905  12/09/2011 12:50       2.10     12680.0          France    Europe   
541906  12/09/2011 12:50       4.15     12680.0          France    Europe   
541907  12/09/2011 12:50       4.15     12680.0          France    Europe   
541908  12/09/2011 12:50       4.95     12680.0          France    Europe   

                 date_time  
0      2010-12-01 08:26:00  
1      2010-12-01 08:26:00  
2      2010-12-01 08:26:00  
3      2010-12-01 08:26:00  
4      2010-12-01 08:26:00  
...                    ...  
541904 2011-12-09 12:50:00  
541905 2011-12-09 12:50:00  
541906 2011-12-09 12:50:00  
541907 2011-12-09 12:50:00  
541908 2011-12-09 12:50:00  

[531283 rows x 10 columns]

# Adding Time Period

In [50]:
data_dadate = data_HW_continent_date_clean1.copy()

In [51]:
data_dadate

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country Continent  \
0       12/01/2010 08:26       2.55     17850.0  United Kingdom    Europe   
1       12/01/2010 08:26       3.39     17850.0  United Kingdom    Europe   
2       12/01/2010 08:26       2.75     17850.0  United Kingdom    Europe   
3       12/01/2010 08:26       3.39     17850.0  United Kingdom    Europe   
4       12/01/2010 08:26       3.39     17850.0  United Kingdom    Europe   
...                  ...        ...         ...             ...       ...   
541904  12/09/2011 12:50       0.85     12680.0          France    Europe   
541905  12/09/2011 12:50       2.10     12680.0          France    Europe   
541906  12/09/2011 12:50       4.15     12680.0          France    Europe   
541907  12/09/2011 12:50       4.15     12680.0          France    Europe   
541908  12/09/2011 12:50       4.95     12680.0          France    Europe   

                 date_time  
0      2010-12-01 08:26:00  
1      2010-12-01 08:26:00  
2      2010-12-01 08:26:00  
3      2010-12-01 08:26:00  
4      2010-12-01 08:26:00  
...                    ...  
541904 2011-12-09 12:50:00  
541905 2011-12-09 12:50:00  
541906 2011-12-09 12:50:00  
541907 2011-12-09 12:50:00  
541908 2011-12-09 12:50:00  

[531283 rows x 10 columns]

## test period time

## Christmas and New Year

In [52]:
# Christmas_period
# Define the date period
start_month = 12  # December
start_day = 1
end_month = 12    # December
end_day = 31

# Check if each date is within the specified period every year
data_dadate['Xmask_NY_Period'] = (
    (data_dadate['date_time'].dt.month > start_month) |
    ((data_dadate['date_time'].dt.month == start_month) & (data_dadate['date_time'].dt.day >= start_day))
) & (
    (data_dadate['date_time'].dt.month < end_month) |
    ((data_dadate['date_time'].dt.month == end_month) & (data_dadate['date_time'].dt.day <= end_day))
)

data_dadate

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country Continent  \
0       12/01/2010 08:26       2.55     17850.0  United Kingdom    Europe   
1       12/01/2010 08:26       3.39     17850.0  United Kingdom    Europe   
2       12/01/2010 08:26       2.75     17850.0  United Kingdom    Europe   
3       12/01/2010 08:26       3.39     17850.0  United Kingdom    Europe   
4       12/01/2010 08:26       3.39     17850.0  United Kingdom    Europe   
...                  ...        ...         ...             ...       ...   
541904  12/09/2011 12:50       0.85     12680.0          France    Europe   
541905  12/09/2011 12:50       2.10     12680.0          France    Europe   
541906  12/09/2011 12:50       4.15     12680.0          France    Europe   
541907  12/09/2011 12:50       4.15     12680.0          France    Europe   
541908  12/09/2011 12:50       4.95     12680.0          France    Europe   

                 date_time  Xmask_NY_Period  
0      2010-12-01 08:26:00             True  
1      2010-12-01 08:26:00             True  
2      2010-12-01 08:26:00             True  
3      2010-12-01 08:26:00             True  
4      2010-12-01 08:26:00             True  
...                    ...              ...  
541904 2011-12-09 12:50:00             True  
541905 2011-12-09 12:50:00             True  
541906 2011-12-09 12:50:00             True  
541907 2011-12-09 12:50:00             True  
541908 2011-12-09 12:50:00             True  

[531283 rows x 11 columns]

## Chinese New Year

In [54]:
# Chinese New Year
# Define the date period
start_month = 1  # January
start_day = 15
end_month = 2    # Febuary
end_day = 15

# Check if each date is within the specified period every year
data_dadate['CNY_Period'] = (
    (data_dadate['date_time'].dt.month > start_month) |
    ((data_dadate['date_time'].dt.month == start_month) & (data_dadate['date_time'].dt.day >= start_day))
) & (
    (data_dadate['date_time'].dt.month < end_month) |
    ((data_dadate['date_time'].dt.month == end_month) & (data_dadate['date_time'].dt.day <= end_day))
)

data_dadate

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country Continent  \
0       12/01/2010 08:26       2.55     17850.0  United Kingdom    Europe   
1       12/01/2010 08:26       3.39     17850.0  United Kingdom    Europe   
2       12/01/2010 08:26       2.75     17850.0  United Kingdom    Europe   
3       12/01/2010 08:26       3.39     17850.0  United Kingdom    Europe   
4       12/01/2010 08:26       3.39     17850.0  United Kingdom    Europe   
...                  ...        ...         ...             ...       ...   
541904  12/09/2011 12:50       0.85     12680.0          France    Europe   
541905  12/09/2011 12:50       2.10     12680.0          France    Europe   
541906  12/09/2011 12:50       4.15     12680.0          France    Europe   
541907  12/09/2011 12:50       4.15     12680.0          France    Europe   
541908  12/09/2011 12:50       4.95     12680.0          France    Europe   

                 date_time  Xmask_NY_Period  CNY_Period  
0      2010-12-01 08:26:00             True       False  
1      2010-12-01 08:26:00             True       False  
2      2010-12-01 08:26:00             True       False  
3      2010-12-01 08:26:00             True       False  
4      2010-12-01 08:26:00             True       False  
...                    ...              ...         ...  
541904 2011-12-09 12:50:00             True       False  
541905 2011-12-09 12:50:00             True       False  
541906 2011-12-09 12:50:00             True       False  
541907 2011-12-09 12:50:00             True       False  
541908 2011-12-09 12:50:00             True       False  

[531283 rows x 12 columns]

In [55]:
data_dadate[data_dadate['CNY_Period']]

InvoiceNo StockCode                          Description  Quantity  \
58345    541238     21115                ROSE CARAVAN DOORSTOP         1   
58346    541238     22957  SET 3 PAPER VINTAGE CHICK PAPER EGG         1   
58347    541238     22758     LARGE PURPLE BABUSHKA NOTEBOOK           1   
58348    541238     20750             RED RETROSPOT MINI CASES         2   
58349    541238     22722    SET OF 6 SPICE TINS PANTRY DESIGN         1   
...         ...       ...                                  ...       ...   
91285    544089    15056N            EDWARDIAN PARASOL NATURAL         3   
91286    544089    84406B       CREAM CUPID HEARTS COAT HANGER         3   
91287    544089     22087             PAPER BUNTING WHITE LACE         3   
91288    544089     22631             CIRCUS PARADE LUNCH BOX          2   
91289    544089     22630                 DOLLY GIRL LUNCH BOX         2   

           InvoiceDate  UnitPrice  CustomerID         Country Continent  \
58345  1/16/2011 10:42       6.75     16746.0  United Kingdom    Europe   
58346  1/16/2011 10:42       2.95     16746.0  United Kingdom    Europe   
58347  1/16/2011 10:42       1.25     16746.0  United Kingdom    Europe   
58348  1/16/2011 10:42       7.95     16746.0  United Kingdom    Europe   
58349  1/16/2011 10:42       3.95     16746.0  United Kingdom    Europe   
...                ...        ...         ...             ...       ...   
91285  2/15/2011 17:51       5.95         NaN  United Kingdom    Europe   
91286  2/15/2011 17:51       4.15         NaN  United Kingdom    Europe   
91287  2/15/2011 17:51       2.95         NaN  United Kingdom    Europe   
91288  2/15/2011 17:51       1.95         NaN  United Kingdom    Europe   
91289  2/15/2011 17:51       1.95         NaN  United Kingdom    Europe   

                date_time  Xmask_NY_Period  CNY_Period  
58345 2011-01-16 10:42:00            False        True  
58346 2011-01-16 10:42:00            False        True  
58347 2011-01-16 10:42:00            False        True  
58348 2011-01-16 10:42:00            False        True  
58349 2011-01-16 10:42:00            False        True  
...                   ...              ...         ...  
91285 2011-02-15 17:51:00            False        True  
91286 2011-02-15 17:51:00            False        True  
91287 2011-02-15 17:51:00            False        True  
91288 2011-02-15 17:51:00            False        True  
91289 2011-02-15 17:51:00            False        True  

[32125 rows x 12 columns]

##Valentine's Day

In [56]:
# Valentine's Day
# Define the date period
start_month = 2  # January
start_day = 14
end_month = 2    # Febuary
end_day = 14

# Check if each date is within the specified period every year
data_dadate['V_Period'] = (
    (data_dadate['date_time'].dt.month > start_month) |
    ((data_dadate['date_time'].dt.month == start_month) & (data_dadate['date_time'].dt.day >= start_day))
) & (
    (data_dadate['date_time'].dt.month < end_month) |
    ((data_dadate['date_time'].dt.month == end_month) & (data_dadate['date_time'].dt.day <= end_day))
)

data_dadate

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country Continent  \
0       12/01/2010 08:26       2.55     17850.0  United Kingdom    Europe   
1       12/01/2010 08:26       3.39     17850.0  United Kingdom    Europe   
2       12/01/2010 08:26       2.75     17850.0  United Kingdom    Europe   
3       12/01/2010 08:26       3.39     17850.0  United Kingdom    Europe   
4       12/01/2010 08:26       3.39     17850.0  United Kingdom    Europe   
...                  ...        ...         ...             ...       ...   
541904  12/09/2011 12:50       0.85     12680.0          France    Europe   
541905  12/09/2011 12:50       2.10     12680.0          France    Europe   
541906  12/09/2011 12:50       4.15     12680.0          France    Europe   
541907  12/09/2011 12:50       4.15     12680.0          France    Europe   
541908  12/09/2011 12:50       4.95     12680.0          France    Europe   

                 date_time  Xmask_NY_Period  CNY_Period  V_Period  
0      2010-12-01 08:26:00             True       False     False  
1      2010-12-01 08:26:00             True       False     False  
2      2010-12-01 08:26:00             True       False     False  
3      2010-12-01 08:26:00             True       False     False  
4      2010-12-01 08:26:00             True       False     False  
...                    ...              ...         ...       ...  
541904 2011-12-09 12:50:00             True       False     False  
541905 2011-12-09 12:50:00             True       False     False  
541906 2011-12-09 12:50:00             True       False     False  
541907 2011-12-09 12:50:00             True       False     False  
541908 2011-12-09 12:50:00             True       False     False  

[531283 rows x 13 columns]

In [57]:
data_dadate[data_dadate['V_Period']]

InvoiceNo StockCode                         Description  Quantity  \
88847    543821     22926      IVORY GIANT GARDEN THERMOMETER         2   
88848    543821     22928     YELLOW GIANT GARDEN THERMOMETER         2   
88849    543821     22925       BLUE GIANT GARDEN THERMOMETER         2   
88850    543821     85175               CACTI T-LIGHT CANDLES        16   
88851    543821     22456     NATURAL SLATE CHALKBOARD LARGE          3   
...         ...       ...                                 ...       ...   
89949    543979     22196        SMALL HEART MEASURING SPOONS         1   
89950    543979     21889              WOODEN BOX OF DOMINOES         1   
89951    543979     85176          SEWING SUSAN 21 NEEDLE SET         1   
89952    543979     37446  MINI CAKE STAND WITH HANGING CAKES         1   
89953    543979     21657     MILK BOTTLE WITH GLASS STOPPER          1   

           InvoiceDate  UnitPrice  CustomerID         Country Continent  \
88847   2/14/2011 9:24       5.95     17667.0  United Kingdom    Europe   
88848   2/14/2011 9:24       5.95     17667.0  United Kingdom    Europe   
88849   2/14/2011 9:24       5.95     17667.0  United Kingdom    Europe   
88850   2/14/2011 9:24       0.42     17667.0  United Kingdom    Europe   
88851   2/14/2011 9:24       4.95     17667.0  United Kingdom    Europe   
...                ...        ...         ...             ...       ...   
89949  2/14/2011 15:43       0.85     14606.0  United Kingdom    Europe   
89950  2/14/2011 15:43       1.25     14606.0  United Kingdom    Europe   
89951  2/14/2011 15:43       0.85     14606.0  United Kingdom    Europe   
89952  2/14/2011 15:43       1.45     14606.0  United Kingdom    Europe   
89953  2/14/2011 15:43       6.95     14606.0  United Kingdom    Europe   

                date_time  Xmask_NY_Period  CNY_Period  V_Period  
88847 2011-02-14 09:24:00            False        True      True  
88848 2011-02-14 09:24:00            False        True      True  
88849 2011-02-14 09:24:00            False        True      True  
88850 2011-02-14 09:24:00            False        True      True  
88851 2011-02-14 09:24:00            False        True      True  
...                   ...              ...         ...       ...  
89949 2011-02-14 15:43:00            False        True      True  
89950 2011-02-14 15:43:00            False        True      True  
89951 2011-02-14 15:43:00            False        True      True  
89952 2011-02-14 15:43:00            False        True      True  
89953 2011-02-14 15:43:00            False        True      True  

[1087 rows x 13 columns]

#HW1 Q1

In [59]:
data_xmas = data_dadate[data_dadate['Xmask_NY_Period']]

In [60]:
transactions_xmas = data_xmas.groupby('InvoiceNo')['StockCode'].apply(list)

In [76]:
min_sup = 0.04

frequentItemsets = list(apriori(transactions_xmas,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'22086', '22910'}), support=0.058752025931928686, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22086', '22910'}), confidence=0.058752025931928686, lift=1.0), OrderedStatistic(items_base=frozenset({'22086'}), items_add=frozenset({'22910'}), confidence=0.5197132616487455, lift=5.938205230319926), OrderedStatistic(items_base=frozenset({'22910'}), items_add=frozenset({'22086'}), confidence=0.6712962962962963, lift=5.938205230319926)])
yes
RelationRecord(items=frozenset({'22111', '22112'}), support=0.04051863857374392, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22111', '22112'}), confidence=0.04051863857374392, lift=1.0), OrderedStatistic(items_base=frozenset({'22111'}), items_add=frozenset({'22112'}), confidence=0.45871559633027525, lift=5.265628333688927), OrderedStatistic(items_base=frozenset({'22112'}), items_add=frozenset({'22111'}), confidence=0.46511627906976744, lift=5.2

RelationRecord(items=frozenset({'22086', '22910'}), support=0.058752025931928686, ordered_statistics=[
  
  OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22086', '22910'}), confidence=0.058752025931928686, lift=1.0),

OrderedStatistic(items_base=frozenset({'22086'}), items_add=frozenset({'22910'}), confidence=0.5197132616487455, lift=5.938205230319926),

 OrderedStatistic(items_base=frozenset({'22910'}), items_add=frozenset({'22086'}), confidence=0.6712962962962963, lift=5.938205230319926)])


In [77]:
set(data_xmas[(data_xmas['StockCode']=='22086')]['Description'])

{"PAPER CHAIN KIT 50'S CHRISTMAS "}

In [78]:
set(data_xmas[(data_xmas['StockCode']=='22910')]['Description'])

{'PAPER CHAIN KIT VINTAGE CHRISTMAS'}

In [79]:
set(data_xmas[(data_xmas['StockCode']=='22835')]['Description'])

{'HOT WATER BOTTLE I AM SO POORLY'}

In [80]:
set(data_xmas[(data_xmas['StockCode']=='22112')]['Description'])

{'CHOCOLATE HOT WATER BOTTLE'}

In [81]:
set(data_xmas[(data_xmas['StockCode']=='22726')]['Description'])

{'ALARM CLOCK BAKELIKE GREEN'}

In [82]:
set(data_xmas[(data_xmas['StockCode']=='22727')]['Description'])

{'ALARM CLOCK BAKELIKE RED '}

In [69]:
set(data_xmas[(data_xmas['StockCode']=='84029E')]['Description'])

{'RED WOOLLY HOTTIE WHITE HEART.'}

In [ ]:
data[data['Country'] == 'Unspecified']

InvoiceNo StockCode                          Description  Quantity  \
152712    549687     20685                DOORMAT RED RETROSPOT         2   
152713    549687     22691              DOORMAT WELCOME SUNRISE         2   
152714    549687     48116           DOORMAT MULTICOLOUR STRIPE         2   
152715    549687     21213          PACK OF 72 SKULL CAKE CASES        24   
152716    549687     21977   PACK OF 60 PINK PAISLEY CAKE CASES        24   
...          ...       ...                                  ...       ...   
498778    578539     22560           TRADITIONAL MODELLING CLAY        24   
498779    578539     23570     TRADITIONAL PICK UP STICKS GAME         12   
498780    578539     23571        TRADITIONAL NAUGHTS & CROSSES        12   
498781    578539     84992       72 SWEETHEART FAIRY CAKE CASES        24   
498782    578539     23309  SET OF 60 I LOVE LONDON CAKE CASES         24   

             InvoiceDate  UnitPrice  CustomerID      Country  
152712  04/11/2011 13:29       7.95     12363.0  Unspecified  
152713  04/11/2011 13:29       7.95     12363.0  Unspecified  
152714  04/11/2011 13:29       7.95     12363.0  Unspecified  
152715  04/11/2011 13:29       0.55     12363.0  Unspecified  
152716  04/11/2011 13:29       0.55     12363.0  Unspecified  
...                  ...        ...         ...          ...  
498778  11/24/2011 14:55       1.25         NaN  Unspecified  
498779  11/24/2011 14:55       1.25         NaN  Unspecified  
498780  11/24/2011 14:55       1.65         NaN  Unspecified  
498781  11/24/2011 14:55       0.55         NaN  Unspecified  
498782  11/24/2011 14:55       0.55         NaN  Unspecified  

[446 rows x 8 columns]

# Q3 ของช่วงเวลา ตรุษจีน ใน ทวีป Asia (15 jan - 15 feb)

In [131]:
data_cny_asia = data_dadate[data_dadate['CNY_Period'] & (data_dadate["Continent"] == "Asia") ]

In [132]:
data_cny_asia

InvoiceNo StockCode                          Description  Quantity  \
69623    541975     22620          4 TRADITIONAL SPINNING TOPS        60   
69624    541975    85183B  CHARLIE & LOLA WASTEPAPER BIN FLORA         6   
69625    541975    85183A   CHARLIE & LOLA WASTEPAPER BIN BLUE         6   
69626    541975     16236                 KITTY PENCIL ERASERS        28   
69627    541975     21058               PARTY INVITES WOODLAND        12   
...         ...       ...                                  ...       ...   
89595    543911    47590B          PINK HAPPY BIRTHDAY BUNTING        12   
89596    543911     22117     METAL SIGN HER DINNER IS SERVED         12   
89597    543911     82567            AIRLINE LOUNGE,METAL SIGN        12   
89598    543911     21910                  WAY OUT METAL SIGN         12   
89599    543911     21911                   GARDEN METAL SIGN         12   

           InvoiceDate  UnitPrice  CustomerID               Country Continent  \
69623  1/24/2011 14:24       1.25         NaN             Hong Kong      Asia   
69624  1/24/2011 14:24       3.75         NaN             Hong Kong      Asia   
69625  1/24/2011 14:24       3.75         NaN             Hong Kong      Asia   
69626  1/24/2011 14:24       0.21         NaN             Hong Kong      Asia   
69627  1/24/2011 14:24       0.85         NaN             Hong Kong      Asia   
...                ...        ...         ...                   ...       ...   
89595  2/14/2011 12:46       5.45     17829.0  United Arab Emirates      Asia   
89596  2/14/2011 12:46       2.95     17829.0  United Arab Emirates      Asia   
89597  2/14/2011 12:46       2.10     17829.0  United Arab Emirates      Asia   
89598  2/14/2011 12:46       1.65     17829.0  United Arab Emirates      Asia   
89599  2/14/2011 12:46       1.65     17829.0  United Arab Emirates      Asia   

                date_time  Xmask_NY_Period  CNY_Period  V_Period  
69623 2011-01-24 14:24:00            False        True     False  
69624 2011-01-24 14:24:00            False        True     False  
69625 2011-01-24 14:24:00            False        True     False  
69626 2011-01-24 14:24:00            False        True     False  
69627 2011-01-24 14:24:00            False        True     False  
...                   ...              ...         ...       ...  
89595 2011-02-14 12:46:00            False        True      True  
89596 2011-02-14 12:46:00            False        True      True  
89597 2011-02-14 12:46:00            False        True      True  
89598 2011-02-14 12:46:00            False        True      True  
89599 2011-02-14 12:46:00            False        True      True  

[273 rows x 13 columns]

In [133]:
transactions_cny = data_cny_asia.groupby('InvoiceNo')['StockCode'].apply(list)

In [135]:
transactions_cny

InvoiceNo
541975    [22620, 85183B, 85183A, 16236, 21058, 21967, 2...
542102    [21519, 22697, 22699, 22343, 22344, 22345, 223...
542276    [82551, 21165, 82600, 21754, 21756, 82482, 824...
543179    [22407, 22406, 22405, 21745, 22496, 22495, 224...
543518    [22759, 22662, 22661, 21717, 22957, 22861, 222...
543911    [21485, 48138, 85150, 82552, 82551, 82583, 224...
Name: StockCode, dtype: object

In [151]:
min_sup = 0.3

frequentItemsets = list(apriori(transactions_cny,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>=3:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'22202', '22203', '20682'}), support=0.3333333333333333, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22202', '22203', '20682'}), confidence=0.3333333333333333, lift=1.0), OrderedStatistic(items_base=frozenset({'20682'}), items_add=frozenset({'22202', '22203'}), confidence=1.0, lift=3.0), OrderedStatistic(items_base=frozenset({'22202'}), items_add=frozenset({'22203', '20682'}), confidence=1.0, lift=3.0), OrderedStatistic(items_base=frozenset({'22203'}), items_add=frozenset({'22202', '20682'}), confidence=1.0, lift=3.0), OrderedStatistic(items_base=frozenset({'22202', '20682'}), items_add=frozenset({'22203'}), confidence=1.0, lift=3.0), OrderedStatistic(items_base=frozenset({'22203', '20682'}), items_add=frozenset({'22202'}), confidence=1.0, lift=3.0), OrderedStatistic(items_base=frozenset({'22202', '22203'}), items_add=frozenset({'20682'}), confidence=1.0, lift=3.0)])
yes
RelationRecord(items=frozenset({'22202', 

# Q4 ช่วงเวลา Christmas & New years ในทวีป Europe (1 Dec - 31 Dec)

In [154]:
data_xmas_eu = data_dadate[data_dadate['Xmask_NY_Period'] & (data_dadate["Continent"]=="Europe")]

In [155]:
data_xmas_eu

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country Continent  \
0       12/01/2010 08:26       2.55     17850.0  United Kingdom    Europe   
1       12/01/2010 08:26       3.39     17850.0  United Kingdom    Europe   
2       12/01/2010 08:26       2.75     17850.0  United Kingdom    Europe   
3       12/01/2010 08:26       3.39     17850.0  United Kingdom    Europe   
4       12/01/2010 08:26       3.39     17850.0  United Kingdom    Europe   
...                  ...        ...         ...             ...       ...   
541904  12/09/2011 12:50       0.85     12680.0          France    Europe   
541905  12/09/2011 12:50       2.10     12680.0          France    Europe   
541906  12/09/2011 12:50       4.15     12680.0          France    Europe   
541907  12/09/2011 12:50       4.15     12680.0          France    Europe   
541908  12/09/2011 12:50       4.95     12680.0          France    Europe   

                 date_time  Xmask_NY_Period  CNY_Period  V_Period  
0      2010-12-01 08:26:00             True       False     False  
1      2010-12-01 08:26:00             True       False     False  
2      2010-12-01 08:26:00             True       False     False  
3      2010-12-01 08:26:00             True       False     False  
4      2010-12-01 08:26:00             True       False     False  
...                    ...              ...         ...       ...  
541904 2011-12-09 12:50:00             True       False     False  
541905 2011-12-09 12:50:00             True       False     False  
541906 2011-12-09 12:50:00             True       False     False  
541907 2011-12-09 12:50:00             True       False     False  
541908 2011-12-09 12:50:00             True       False     False  

[66688 rows x 13 columns]

In [156]:
transactions_xmas_eu = data_xmas_eu.groupby('InvoiceNo')['StockCode'].apply(list)

In [173]:
min_sup = 0.04

frequentItemsets = list(apriori(transactions_xmas_eu,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'22086', '22910'}), support=0.05896705978039853, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22086', '22910'}), confidence=0.05896705978039853, lift=1.0), OrderedStatistic(items_base=frozenset({'22086'}), items_add=frozenset({'22910'}), confidence=0.5197132616487454, lift=5.91655051108456), OrderedStatistic(items_base=frozenset({'22910'}), items_add=frozenset({'22086'}), confidence=0.6712962962962963, lift=5.916550511084561)])
yes
RelationRecord(items=frozenset({'22111', '22112'}), support=0.040666937779585195, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22111', '22112'}), confidence=0.040666937779585195, lift=1.0), OrderedStatistic(items_base=frozenset({'22111'}), items_add=frozenset({'22112'}), confidence=0.4587155963302752, lift=5.2464262854704495), OrderedStatistic(items_base=frozenset({'22112'}), items_add=frozenset({'22111'}), confidence=0.46511627906976744, lift=5.24

# Q5 ของวันวาเลนไทน์

In [182]:
data_v = data_dadate[data_dadate['V_Period']]


In [183]:
data_v

InvoiceNo StockCode                         Description  Quantity  \
88847    543821     22926      IVORY GIANT GARDEN THERMOMETER         2   
88848    543821     22928     YELLOW GIANT GARDEN THERMOMETER         2   
88849    543821     22925       BLUE GIANT GARDEN THERMOMETER         2   
88850    543821     85175               CACTI T-LIGHT CANDLES        16   
88851    543821     22456     NATURAL SLATE CHALKBOARD LARGE          3   
...         ...       ...                                 ...       ...   
89949    543979     22196        SMALL HEART MEASURING SPOONS         1   
89950    543979     21889              WOODEN BOX OF DOMINOES         1   
89951    543979     85176          SEWING SUSAN 21 NEEDLE SET         1   
89952    543979     37446  MINI CAKE STAND WITH HANGING CAKES         1   
89953    543979     21657     MILK BOTTLE WITH GLASS STOPPER          1   

           InvoiceDate  UnitPrice  CustomerID         Country Continent  \
88847   2/14/2011 9:24       5.95     17667.0  United Kingdom    Europe   
88848   2/14/2011 9:24       5.95     17667.0  United Kingdom    Europe   
88849   2/14/2011 9:24       5.95     17667.0  United Kingdom    Europe   
88850   2/14/2011 9:24       0.42     17667.0  United Kingdom    Europe   
88851   2/14/2011 9:24       4.95     17667.0  United Kingdom    Europe   
...                ...        ...         ...             ...       ...   
89949  2/14/2011 15:43       0.85     14606.0  United Kingdom    Europe   
89950  2/14/2011 15:43       1.25     14606.0  United Kingdom    Europe   
89951  2/14/2011 15:43       0.85     14606.0  United Kingdom    Europe   
89952  2/14/2011 15:43       1.45     14606.0  United Kingdom    Europe   
89953  2/14/2011 15:43       6.95     14606.0  United Kingdom    Europe   

                date_time  Xmask_NY_Period  CNY_Period  V_Period  
88847 2011-02-14 09:24:00            False        True      True  
88848 2011-02-14 09:24:00            False        True      True  
88849 2011-02-14 09:24:00            False        True      True  
88850 2011-02-14 09:24:00            False        True      True  
88851 2011-02-14 09:24:00            False        True      True  
...                   ...              ...         ...       ...  
89949 2011-02-14 15:43:00            False        True      True  
89950 2011-02-14 15:43:00            False        True      True  
89951 2011-02-14 15:43:00            False        True      True  
89952 2011-02-14 15:43:00            False        True      True  
89953 2011-02-14 15:43:00            False        True      True  

[1087 rows x 13 columns]

In [185]:
transactions_v = data_v.groupby('InvoiceNo')['StockCode'].apply(list)
transactions_v

InvoiceNo
543821    [22926, 22928, 22925, 85175, 22456, 22854, 227...
543822    [22139, 20973, 22585, 22960, 84380, 84378, 229...
543823                         [22138, 21658, 22567, 20914]
543824    [22745, 22746, 20971, 20972, 16225, 22748, 227...
543825    [22926, 22782, 22411, 20914, 22381, 20681, 22690]
543826    [90104, 90146, 90175A, 90181B, 90214J, 90214S,...
543828    [22723, 22722, 22720, 22899, 22567, 84987, 228...
543829    [21238, 21239, 21240, 84997B, 84997D, 84997C, ...
543831    [84519A, 20711, 20713, 85099F, 22385, 21929, 2...
543832    [84879, 22077, 22855, 21899, 21900, 21901, 219...
543833    [22788, 22957, 21621, 22722, 22384, 21791, 217...
543836    [84987, 84991, 84988, 84992, 84378, 22965, 229...
543839                               [22207, 21731, 47590A]
543842                                              [22693]
543852    [22355, 21931, 22356, 85123A, 85175, 82484, 82...
543899    [C2, 21340, 21792, 22503, 15056N, 22630, 22983...
543900    [22355, 20719, 20723

In [191]:
min_sup = 0.1

frequentItemsets = list(apriori(transactions_v,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'22720', '22722'}), support=0.11904761904761904, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22720', '22722'}), confidence=0.11904761904761904, lift=1.0), OrderedStatistic(items_base=frozenset({'22720'}), items_add=frozenset({'22722'}), confidence=0.7142857142857143, lift=5.0), OrderedStatistic(items_base=frozenset({'22722'}), items_add=frozenset({'22720'}), confidence=0.8333333333333334, lift=5.000000000000001)])
yes
RelationRecord(items=frozenset({'22720', '22969'}), support=0.11904761904761904, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22720', '22969'}), confidence=0.11904761904761904, lift=1.0), OrderedStatistic(items_base=frozenset({'22720'}), items_add=frozenset({'22969'}), confidence=0.7142857142857143, lift=6.000000000000001), OrderedStatistic(items_base=frozenset({'22969'}), items_add=frozenset({'22720'}), confidence=1.0, lift=6.0)])
